### PR TITLE
[Feature branch] HTTP Dataset source

### PR DESCRIPTION
**Pull request by dbczumar**
## What changes are proposed in this pull request?

Introduce an HTTP Dataset source to MLflow Tracking. Also make some other fixes to existing dataset sources and add test coverage.

## How is this patch tested?

- [ ] Existing unit/integration tests
- [x] New unit/integration tests
- [ ] Manual tests (describe details, including test results, below)

## Does this PR change the documentation?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: tests/data/test_delta_dataset_source.py [hunk 1]
```diff
@@ -1,6 +1,8 @@
+import json
```
Comment (dbczumar) [inline tests/data/test_delta_dataset_source.py:1]: Improve dataset source coverage while we're here.

File: tests/data/test_http_dataset_source.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+def test_source_load(tmp_path):
+    source1 = HTTPDatasetSource(
+        "https://raw.githubusercontent.com/mlflow/mlflow/master/tests/datasets/winequality-red.csv"
+    )
+
+    loaded1 = source1.load()
+    parsed1 = pd.read_csv(loaded1, sep=";")
+    assert "fixed acidity" in parsed1.columns
+    assert len(parsed1) > 10
```
Comment (prithvikannan) [inline tests/data/test_http_dataset_source.py:56]: what is the purpose of this assertion?

File: tests/data/test_http_dataset_source.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    loaded3 = source2.load(dst_path=tmp_path)
+    assert loaded3 == str(tmp_path / "winequality-red.csv")
+    parsed3 = pd.read_csv(loaded3, sep=";")
+    assert "fixed acidity" in parsed3.columns
+    assert len(parsed1) > 10
+
+    source3 = HTTPDatasetSource("https://github.com/")
+    loaded4 = source3.load()
+    assert os.path.exists(loaded4)
+    assert os.path.basename(loaded4) == "dataset_source"
```
Comment (prithvikannan) [inline tests/data/test_http_dataset_source.py:76]: interesting test. we have essentially created a "wget" like function with `HTTPDatasetSource.load()` :)

File: mlflow/data/http_dataset_source.py [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+        augmented_raise_for_status(resp)
+
+        path = urlparse(self.url).path
+        content_disposition = resp.headers.get("Content-Disposition")
+        if content_disposition is not None and len(content_disposition) > 0:
+            basename = content_disposition
+        elif path is not None and len(posixpath.basename(path)) > 0:
+            basename = posixpath.basename(path)
+        else:
+            basename = "dataset_source"
```
Comment (prithvikannan) [inline mlflow/data/http_dataset_source.py:62]: can we add a test case for this logic?

File: mlflow/data/http_dataset_source.py [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+                         dataset source. If the directory does not exist, it is created. If
+                         unspecified, the dataset source is downloaded to a new uniquely-named
+                         directory on the local filesystem.
+        :return: The path to the downloaded dataset source on the local filesystem.
+        """
+        resp = cloud_storage_http_request(
+            method="GET",
+            url=self.url,
+        )
+        augmented_raise_for_status(resp)
```
Comment (prithvikannan) [inline mlflow/data/http_dataset_source.py:53]: can we also add a test for a well formed URL that doesn't exist?

**Review by prithvikannan — COMMENTED** [review]
This is great! I think we should add a bit more test coverage around the `load()` function.

File: mlflow/data/http_dataset_source.py (hunk 5 above)
Comment (dbczumar) [reply mlflow/data/http_dataset_source.py:54]: Added a case for this to `test_source_load`

File: tests/data/test_http_dataset_source.py (hunk 2 above)
Comment (dbczumar) [reply tests/data/test_http_dataset_source.py:60]: Added an inline comment. It verifies that we actually downloaded some dataset rows.

File: tests/data/test_http_dataset_source.py (hunk 3 above)
Comment (dbczumar) [reply tests/data/test_http_dataset_source.py:82]: Exactly :)

**Review by dbczumar — COMMENTED** [review]
@prithvikannan Thanks so much for the detailed review. I've addressed your comments. Can you take another look?

**Review by prithvikannan — APPROVED** [review]
Looks good @dbczumar ! Thanks for making all the fixes here.

**Pull request merged**